{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "781923\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "import pathlib\n",
    "import pickle\n",
    "import collections\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import networkx as nx\n",
    "from Bio import SeqIO, Align\n",
    "from matplotlib import colormaps\n",
    "from scipy.stats import pearsonr\n",
    "from datetime import datetime\n",
    "from sklearn.decomposition import PCA\n",
    "#adding portek source directory to sys path before importing\n",
    "portek_path = \"../portek\"\n",
    "sys.path.insert(0,portek_path)\n",
    "import portek\n",
    "\n",
    "pd.options.mode.copy_on_write = True\n",
    "print(os.getpid())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PORT-EK parameters\n",
    "k = 11  # This is the length of k-mers\n",
    "c = 0.8  # This is the conservation thershold used in k-mer rarity filter\n",
    "m = 1  # This is the maximum number of mismatches allowed when re-examining rare k-mers\n",
    "min_RMSE = 0.5  # This is the RMSE threshold used to select enriched k-mers\n",
    "\n",
    "# Relative path to k-mer indices\n",
    "INPUT_PATH = \"../output/test\"\n",
    "\n",
    "# Data set specific definitions of k-mer type, column names, reference gene and protein mapping, and colors for plots.\n",
    "SAMPLE_GROUPS = [\"MD\",\"N\",\"O\"]\n",
    "MODE = \"all_vs_all\"\n",
    "GROUP_OF_INTEREST = \"MD\"\n",
    "\n",
    "GENES = {\n",
    "    \"5' LTR\": (1, 634),\n",
    "    \"gag\": (790, 2292),\n",
    "    \"pol\": (2358, 5096),\n",
    "    \"vif\": (5041, 5619),\n",
    "    \"vpr\": (5559, 5795),\n",
    "    \"tat_1ex\": (5831, 6045),\n",
    "    \"tat_2ex\": (8379, 8653),\n",
    "    \"rev_1ex\": (5970, 6045),\n",
    "    \"rev_2ex\": (8379, 8653),\n",
    "    \"env\": (6225, 8795),\n",
    "    \"nef\": (8797, 9168),\n",
    "    \"3' LTR\": (9086, 9719),\n",
    "}\n",
    "\n",
    "REF_SEQ = SeqIO.read(\"../input/HXB2.fasta\", format=\"fasta\").seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "FREQ_COLS = [f\"{group}_freq\" for group in SAMPLE_GROUPS]\n",
    "AVG_COLS = [f\"{group}_avg\" for group in SAMPLE_GROUPS]\n",
    "aligner = Align.PairwiseAligner(\n",
    "    scoring=\"megablast\",\n",
    "    mode=\"local\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Imported 276056 kmers and 293 samples.\n"
     ]
    }
   ],
   "source": [
    "# First read all sample labels and k-mer sequences and construct an empty matrix. It requires significantly less memory than growing the matrix one-by-one.\n",
    "kmer_set = set()\n",
    "sample_list = []\n",
    "kmer_set_in_path = pathlib.Path(INPUT_PATH).glob(f\"*{k}mer_set.pkl\")\n",
    "sample_list_in_path = pathlib.Path(INPUT_PATH).glob(\"*sample_list.pkl\")\n",
    "\n",
    "for filename in kmer_set_in_path:\n",
    "    with open(filename, mode=\"rb\") as in_file:\n",
    "        partial_set = pickle.load(in_file)\n",
    "    kmer_set.update(partial_set)\n",
    "\n",
    "for filename in sample_list_in_path:\n",
    "    with open(filename, mode=\"rb\") as in_file:\n",
    "        partial_list = pickle.load(in_file)\n",
    "    group = filename.stem.split(\"_\")[0]\n",
    "    partial_list = [f\"{group}_{sample_name}\" for sample_name in partial_list]\n",
    "    sample_list.extend(partial_list)\n",
    "sample_list.sort()\n",
    "\n",
    "all_kmer_matrix = pd.DataFrame(\n",
    "    0, index=list(kmer_set), columns=sample_list, dtype=\"uint8\"\n",
    ")\n",
    "\n",
    "group_sample_idx = {f\"{group}\":[sample for sample in sample_list if sample.split(\"_\")[0] == f\"{group}\"] for group in SAMPLE_GROUPS}\n",
    "\n",
    "print(f\"\\nImported {len(kmer_set)} kmers and {len(sample_list)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "293 of 293 indices done.\n",
      "5697 common k-mers remaining after filtering at a threshold of 0.8.\n"
     ]
    }
   ],
   "source": [
    "# Next, fill the table with k-mer counts.\n",
    "counter = 1\n",
    "tot_files = len(sample_list)\n",
    "in_path = pathlib.Path(f\"{INPUT_PATH}/{k}mer_indices\").glob(\"*_count.pkl\")\n",
    "\n",
    "for filename in in_path:\n",
    "    with open(filename, mode=\"rb\") as in_file:\n",
    "        temp_dict = pickle.load(in_file)\n",
    "    sample_name = \"_\".join(filename.stem.split(\"_\")[:-1])\n",
    "    count_dict = {f\"{sample_name}\": temp_dict.values()}\n",
    "    temp_df = pd.DataFrame(count_dict, index=temp_dict.keys(), dtype=\"uint8\")\n",
    "    all_kmer_matrix.update(temp_df)\n",
    "    print(\n",
    "        f\"{counter} of {tot_files} indices done.\",\n",
    "        end=\"\\r\",\n",
    "        flush=True,\n",
    "    )\n",
    "    counter += 1\n",
    "\n",
    "# Decode k-mer sequences\n",
    "all_kmer_matrix.index = all_kmer_matrix.index.map(lambda id: portek.decode_kmer(id,k))\n",
    "\n",
    "# Construct a temporary binary count matrix, i.e. one that shows if a k-mer is present in sequence, without regards to actual count.\n",
    "# Calculate k-mer frequencies and average counts in host groups.\n",
    "bin_kmer_matrix = all_kmer_matrix > 0\n",
    "for group in SAMPLE_GROUPS:\n",
    "    all_kmer_matrix[f\"{group}_freq\"] = bin_kmer_matrix.loc[:, group_sample_idx[group]].mean(axis=1)\n",
    "    all_kmer_matrix[f\"{group}_avg\"] = all_kmer_matrix.loc[:, group_sample_idx[group]].mean(axis=1)\n",
    "del bin_kmer_matrix\n",
    "\n",
    "# Remove polyA, as its presence and count is mostly dependant on sequencing quality not viral variant.\n",
    "if k*\"A\" in all_kmer_matrix.index:\n",
    "    all_kmer_matrix = all_kmer_matrix.drop(k*\"A\")\n",
    "\n",
    "# Apply rarity filter.\n",
    "common_kmer_matrix = portek.filter_kmers(\n",
    "    all_kmer_matrix, freq_cols=FREQ_COLS, cons_thr=c\n",
    ")\n",
    "\n",
    "print(\n",
    "    f\"\\n{len(common_kmer_matrix)} common k-mers remaining after filtering at a threshold of {c}.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All_vs_All vs One_vs_Rest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "if MODE == \"all_vs_all\":\n",
    "    common_kmer_matrix[\"seq\"] = common_kmer_matrix.index\n",
    "    ERR_COLS = []\n",
    "    P_COLS = []\n",
    "    for j in range(1, len(SAMPLE_GROUPS)):\n",
    "        for i in range(j):\n",
    "            err_name = f\"{SAMPLE_GROUPS[i]}-{SAMPLE_GROUPS[j]}_err\"\n",
    "            p_name = f\"{SAMPLE_GROUPS[i]}-{SAMPLE_GROUPS[j]}_p-value\"\n",
    "            ERR_COLS.append(err_name)\n",
    "            P_COLS.append(p_name)\n",
    "            common_kmer_matrix[err_name] = (\n",
    "                common_kmer_matrix[f\"{SAMPLE_GROUPS[i]}_avg\"]\n",
    "                - common_kmer_matrix[f\"{SAMPLE_GROUPS[j]}_avg\"]\n",
    "            )\n",
    "            common_kmer_matrix[p_name] = common_kmer_matrix[\"seq\"].apply(\n",
    "                portek.calc_kmer_pvalue,\n",
    "                args=(\n",
    "                    group_sample_idx[SAMPLE_GROUPS[i]],\n",
    "                    group_sample_idx[SAMPLE_GROUPS[j]],\n",
    "                    common_kmer_matrix,\n",
    "                ),\n",
    "            )\n",
    "            common_kmer_matrix[f\"-log10_{p_name}\"] = -np.log10(common_kmer_matrix[p_name])\n",
    "    common_kmer_matrix[\"RMSE\"] = np.sqrt(((common_kmer_matrix[ERR_COLS]) ** 2).mean(axis=1))\n",
    "    common_kmer_matrix = common_kmer_matrix.sort_values(\"RMSE\", ascending=False)\n",
    "    common_kmer_matrix = common_kmer_matrix.drop(\"seq\", axis=1)\n",
    "    common_kmer_matrix[\"group\"] = common_kmer_matrix.apply(\n",
    "        portek.assign_kmer_group_ava, p_cols = P_COLS, avg_cols = AVG_COLS,axis=1\n",
    "    )\n",
    "    common_kmer_matrix[\"exclusivity\"] = common_kmer_matrix.apply(\n",
    "        portek.check_exclusivity, avg_cols = AVG_COLS, axis=1\n",
    "    )\n",
    "    common_kmer_stat_matrix = common_kmer_matrix.drop(sample_list, axis=1)\n",
    "\n",
    "else:\n",
    "    common_kmer_matrix[\"seq\"] = common_kmer_matrix.index\n",
    "    ERR_COLS = []\n",
    "    P_COLS = []\n",
    "    control_groups = SAMPLE_GROUPS.copy()\n",
    "    control_groups.remove(GROUP_OF_INTEREST)\n",
    "    for j in range(len(control_groups)):\n",
    "        err_name = f\"{GROUP_OF_INTEREST}-{control_groups[j]}_err\"\n",
    "        p_name = f\"{GROUP_OF_INTEREST}-{control_groups[j]}_p-value\"\n",
    "        ERR_COLS.append(err_name)\n",
    "        P_COLS.append(p_name)\n",
    "        common_kmer_matrix[err_name] = (\n",
    "            common_kmer_matrix[f\"{GROUP_OF_INTEREST}_avg\"]\n",
    "            - common_kmer_matrix[f\"{control_groups[j]}_avg\"]\n",
    "        )\n",
    "        common_kmer_matrix[p_name] = common_kmer_matrix[\"seq\"].apply(\n",
    "            portek.calc_kmer_pvalue,\n",
    "            args=(\n",
    "                group_sample_idx[GROUP_OF_INTEREST],\n",
    "                group_sample_idx[control_groups[j]],\n",
    "                common_kmer_matrix,\n",
    "            ),\n",
    "        )\n",
    "        common_kmer_matrix[f\"-log10_{p_name}\"] = -np.log10(common_kmer_matrix[p_name])\n",
    "    common_kmer_matrix[\"RMSE\"] = np.sqrt(((common_kmer_matrix[ERR_COLS]) ** 2).mean(axis=1))\n",
    "    common_kmer_matrix = common_kmer_matrix.sort_values(\"RMSE\", ascending=False)\n",
    "    common_kmer_matrix = common_kmer_matrix.drop(\"seq\", axis=1)\n",
    "    common_kmer_matrix[\"group\"] = common_kmer_matrix.apply(\n",
    "        portek.assign_kmer_group, goi = GROUP_OF_INTEREST, p_cols = P_COLS, err_cols = ERR_COLS,axis=1\n",
    "    )\n",
    "    common_kmer_matrix[\"exclusivity\"] = common_kmer_matrix.apply(\n",
    "        portek.check_exclusivity, avg_cols = AVG_COLS, axis=1\n",
    "    )\n",
    "    common_kmer_stat_matrix = common_kmer_matrix.drop(sample_list, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "group\n",
       "N_enriched         2396\n",
       "MD_enriched        1409\n",
       "not_significant     981\n",
       "O_enriched          911\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_kmer_stat_matrix[\"group\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "group\n",
       "not_significant     3805\n",
       "MD_enriched         1409\n",
       "control_enriched     483\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_kmer_stat_matrix_ovr[\"group\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MD_freq</th>\n",
       "      <th>MD_avg</th>\n",
       "      <th>N_freq</th>\n",
       "      <th>N_avg</th>\n",
       "      <th>O_freq</th>\n",
       "      <th>O_avg</th>\n",
       "      <th>MD-N_err</th>\n",
       "      <th>MD-N_p-value</th>\n",
       "      <th>-log10_MD-N_p-value</th>\n",
       "      <th>MD-O_err</th>\n",
       "      <th>MD-O_p-value</th>\n",
       "      <th>-log10_MD-O_p-value</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>group</th>\n",
       "      <th>exclusivity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACAGGAGCAGA</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.675556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>1.675556</td>\n",
       "      <td>4.087578e-14</td>\n",
       "      <td>13.388534</td>\n",
       "      <td>1.640468</td>\n",
       "      <td>2.630821e-35</td>\n",
       "      <td>34.579909</td>\n",
       "      <td>1.658105</td>\n",
       "      <td>MD_enriched</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAAAAATAGA</th>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>1.363636</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.842105</td>\n",
       "      <td>-1.336970</td>\n",
       "      <td>5.165618e-07</td>\n",
       "      <td>6.286878</td>\n",
       "      <td>-1.815439</td>\n",
       "      <td>1.008934e-12</td>\n",
       "      <td>11.996137</td>\n",
       "      <td>1.594256</td>\n",
       "      <td>control_enriched</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAATTGGAGA</th>\n",
       "      <td>0.937778</td>\n",
       "      <td>1.533333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.533333</td>\n",
       "      <td>2.502012e-08</td>\n",
       "      <td>7.601711</td>\n",
       "      <td>1.533333</td>\n",
       "      <td>5.159604e-28</td>\n",
       "      <td>27.287384</td>\n",
       "      <td>1.533333</td>\n",
       "      <td>MD_enriched</td>\n",
       "      <td>exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGAATGTATA</th>\n",
       "      <td>0.955556</td>\n",
       "      <td>1.502222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.502222</td>\n",
       "      <td>6.066930e-10</td>\n",
       "      <td>9.217031</td>\n",
       "      <td>1.502222</td>\n",
       "      <td>3.279253e-33</td>\n",
       "      <td>32.484225</td>\n",
       "      <td>1.502222</td>\n",
       "      <td>MD_enriched</td>\n",
       "      <td>exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGACTTTCCA</th>\n",
       "      <td>0.866667</td>\n",
       "      <td>1.017778</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.087719</td>\n",
       "      <td>0.381414</td>\n",
       "      <td>8.714941e-02</td>\n",
       "      <td>1.059736</td>\n",
       "      <td>-2.069942</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.488310</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCAGGAAGAAG</th>\n",
       "      <td>0.991111</td>\n",
       "      <td>0.991111</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.008889</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.008889</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.008889</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGATTGGG</th>\n",
       "      <td>0.991111</td>\n",
       "      <td>0.991111</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>0.982456</td>\n",
       "      <td>-0.008889</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.008655</td>\n",
       "      <td>4.934504e-01</td>\n",
       "      <td>0.306756</td>\n",
       "      <td>0.008773</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATTGGATGAC</th>\n",
       "      <td>0.893333</td>\n",
       "      <td>0.902222</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.912281</td>\n",
       "      <td>0.912281</td>\n",
       "      <td>-0.006869</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.010058</td>\n",
       "      <td>8.097566e-01</td>\n",
       "      <td>0.091645</td>\n",
       "      <td>0.008613</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAATTATGGTA</th>\n",
       "      <td>0.897778</td>\n",
       "      <td>0.897778</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>-0.011313</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.003041</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.008284</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGAAAGGACC</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>not_significant</td>\n",
       "      <td>non-exclusive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5697 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MD_freq    MD_avg    N_freq     N_avg    O_freq     O_avg  \\\n",
       "ACAGGAGCAGA  1.000000  1.675556  0.000000  0.000000  0.035088  0.035088   \n",
       "GAAAAAATAGA  0.026667  0.026667  0.909091  1.363636  1.000000  1.842105   \n",
       "ACAATTGGAGA  0.937778  1.533333  0.000000  0.000000  0.000000  0.000000   \n",
       "AAGAATGTATA  0.955556  1.502222  0.000000  0.000000  0.000000  0.000000   \n",
       "GGGACTTTCCA  0.866667  1.017778  0.636364  0.636364  1.000000  3.087719   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "GCAGGAAGAAG  0.991111  0.991111  1.000000  1.000000  1.000000  1.000000   \n",
       "GGGGGATTGGG  0.991111  0.991111  1.000000  1.000000  0.982456  0.982456   \n",
       "AATTGGATGAC  0.893333  0.902222  0.909091  0.909091  0.912281  0.912281   \n",
       "AAATTATGGTA  0.897778  0.897778  0.909091  0.909091  0.894737  0.894737   \n",
       "TGGAAAGGACC  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   \n",
       "\n",
       "             MD-N_err  MD-N_p-value  -log10_MD-N_p-value  MD-O_err  \\\n",
       "ACAGGAGCAGA  1.675556  4.087578e-14            13.388534  1.640468   \n",
       "GAAAAAATAGA -1.336970  5.165618e-07             6.286878 -1.815439   \n",
       "ACAATTGGAGA  1.533333  2.502012e-08             7.601711  1.533333   \n",
       "AAGAATGTATA  1.502222  6.066930e-10             9.217031  1.502222   \n",
       "GGGACTTTCCA  0.381414  8.714941e-02             1.059736 -2.069942   \n",
       "...               ...           ...                  ...       ...   \n",
       "GCAGGAAGAAG -0.008889  1.000000e+00            -0.000000 -0.008889   \n",
       "GGGGGATTGGG -0.008889  1.000000e+00            -0.000000  0.008655   \n",
       "AATTGGATGAC -0.006869  1.000000e+00            -0.000000 -0.010058   \n",
       "AAATTATGGTA -0.011313  1.000000e+00            -0.000000  0.003041   \n",
       "TGGAAAGGACC  0.000000  1.000000e+00            -0.000000  0.000000   \n",
       "\n",
       "             MD-O_p-value  -log10_MD-O_p-value      RMSE             group  \\\n",
       "ACAGGAGCAGA  2.630821e-35            34.579909  1.658105       MD_enriched   \n",
       "GAAAAAATAGA  1.008934e-12            11.996137  1.594256  control_enriched   \n",
       "ACAATTGGAGA  5.159604e-28            27.287384  1.533333       MD_enriched   \n",
       "AAGAATGTATA  3.279253e-33            32.484225  1.502222       MD_enriched   \n",
       "GGGACTTTCCA  1.000000e+00            -0.000000  1.488310   not_significant   \n",
       "...                   ...                  ...       ...               ...   \n",
       "GCAGGAAGAAG  1.000000e+00            -0.000000  0.008889   not_significant   \n",
       "GGGGGATTGGG  4.934504e-01             0.306756  0.008773   not_significant   \n",
       "AATTGGATGAC  8.097566e-01             0.091645  0.008613   not_significant   \n",
       "AAATTATGGTA  1.000000e+00            -0.000000  0.008284   not_significant   \n",
       "TGGAAAGGACC  1.000000e+00            -0.000000  0.000000   not_significant   \n",
       "\n",
       "               exclusivity  \n",
       "ACAGGAGCAGA  non-exclusive  \n",
       "GAAAAAATAGA  non-exclusive  \n",
       "ACAATTGGAGA      exclusive  \n",
       "AAGAATGTATA      exclusive  \n",
       "GGGACTTTCCA  non-exclusive  \n",
       "...                    ...  \n",
       "GCAGGAAGAAG  non-exclusive  \n",
       "GGGGGATTGGG  non-exclusive  \n",
       "AATTGGATGAC  non-exclusive  \n",
       "AAATTATGGTA  non-exclusive  \n",
       "TGGAAAGGACC  non-exclusive  \n",
       "\n",
       "[5697 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_kmer_stat_matrix_ovr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MD_freq                     0.053333\n",
       "MD_avg                      0.057778\n",
       "N_freq                           0.0\n",
       "N_avg                            0.0\n",
       "O_freq                      0.877193\n",
       "O_avg                       1.473684\n",
       "MD-N_err                    0.057778\n",
       "MD-N_p-value                     1.0\n",
       "-log10_MD-N_p-value             -0.0\n",
       "MD-O_err                   -1.415906\n",
       "MD-O_p-value                     0.0\n",
       "-log10_MD-O_p-value         12.38343\n",
       "N-O_err                    -1.473684\n",
       "N-O_p-value                 0.000127\n",
       "-log10_N-O_p-value          3.897902\n",
       "RMSE                        1.180378\n",
       "group                     O_enriched\n",
       "exclusivity            non-exclusive\n",
       "Name: AGCCCGGGAGC, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "differing_kmers = []\n",
    "for kmer in common_kmer_stat_matrix.index:\n",
    "    if common_kmer_stat_matrix.loc[kmer, \"group\"] != common_kmer_stat_matrix_ovr.loc[kmer, \"group\"]:\n",
    "        differing_kmers.append(kmer)\n",
    "\n",
    "common_kmer_stat_matrix.loc[differing_kmers[14]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MD_freq                       0.053333\n",
       "MD_avg                        0.057778\n",
       "N_freq                             0.0\n",
       "N_avg                              0.0\n",
       "O_freq                        0.877193\n",
       "O_avg                         1.473684\n",
       "MD-N_err                      0.057778\n",
       "MD-N_p-value                       1.0\n",
       "-log10_MD-N_p-value               -0.0\n",
       "MD-O_err                     -1.415906\n",
       "MD-O_p-value                       0.0\n",
       "-log10_MD-O_p-value           12.38343\n",
       "RMSE                           1.00203\n",
       "group                  not_significant\n",
       "exclusivity              non-exclusive\n",
       "Name: AGCCCGGGAGC, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_kmer_stat_matrix_ovr.loc[differing_kmers[14]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
